kubelet: Add preparePod to check pod infra container.

yifan-gu opened the pull request:
In preparePod, it will check the pod infra container, restart it if necessary.
If the pod infra container is restarted, all running containers in the pod will
be killed as well. (https://github.com/GoogleCloudPlatform/kubernetes/issues/5896)

As discussed in https://github.com/GoogleCloudPlatform/kubernetes/pull/6068#issuecomment-87128723

I tried the 3) solution here. The syncPod is now generic with only a specific code path for docker container. Tests might be broken, but want to hear your advices! Thank you!

@vmarmol @dchen1107 @pmorie

vmarmol commented:
@dchen1107 please wait for me to merge :) I'm a little behind because of on-call, but I wanted to think through this. I didn't like the approaches we were taking.

pmorie commented:
Gut reaction is I like where this is headed, still reading though

pmorie commented on pkg/kubelet/kubelet.go:
```diff
@@ -932,12 +929,20 @@ func (kl *Kubelet) pullImage(img string, ref *api.ObjectReference) error {
 	for _, container := range runningPod.Containers {
 		wg.Add(1)
 		go func(container *kubecontainer.Container) {
+			// Call the networking plugin for teardown.
```
A couple thoughts:

First, there is no validation to prevent a user from creating a container with the reserved name.  Should we add one?

Second, this is one of the many call sites where it would be really advantageous to have a convenient way of marking a container as being a system artifact, ie, not present in user config/payloads.  Thinking back to another life time, I remembered that the Java reflection API has a notion called 'synthetic' in the reflection API - a synthetic method is one that is not present in user code but is generated by the compiler -- for example, implied no-arg constructors, serialization constructors, constructors to nested classes, etc.  I think this at some level is the same idea.  So, what about this:

1.  Add a 'Synthetic' field to `Container`
2.  Make it so that a user cannot set the value of this field; ie, it is ignored on create / update in the rest API.
3.  Kubelet or future container runtime can use this field to mark synthetic containers it creates (ie, pod infra container)

Any thoughts @yifan-gu @vmarmol @dchen1107 @thockin ?

yifan-gu replied:
@pmorie I think a bigger question to this is whether we want the kubelet knows those 'synthetic' containers. Should we hide it under the container runtime? Should it even be included in the `runningPod`?
If we just let the container runtime handle the infra container, it might do some killing/restarting out of the kubelet's control.

pmorie replied:
@yifan-gu I think I will file another issue for this, there are a ton of places in the code and in issues where this is hinted at, and it's really orthogonal from the thrust of this PR.

Commit:
kubelet: Add preparePod to check pod infra container.

In preparePod, it will check the pod infra container, restart it if necessary.
If the pod infra container is restarted, all running containers in the pod will
be killed as well.

Commit:
kubelet: Add preparePod to check pod infra container.

Updating unit tests.

yifan-gu commented:
Just fixed all existing tests. Now going to add a few more to exercise the new code path.

Commit:
kubelet: Add preparePod to check pod infra container.

Add a new test to test killing pod infra container when its hash changed.

yifan-gu commented:
I think this is ready for a review @vmarmol :)

yifan-gu commented:
@vmarmol BTW, do you want to discuss the approaches in a hangout maybe?

pmorie commented:
@yifan-gu FWIW this LGTM provisionally... syncPod is getting more grokkable

yifan-gu commented:
@pmorie :) BTW, While refactoring, I was thinking of replacing `kubecontainer.Pod` with `api.Pod`, I think we can do that if we add a `hash` field in the `ContainerStatus`. But we don't need to rush in a hurry, just keep the `kubecontainer.Pod` for now is fine, and more flexible. 
/cc @dchen1107

yifan-gu commented:
@vmarmol @dchen1107 Any thoughts over this yet? :ear:

yifan-gu commented:
@vmarmol Just FYI, if you are going to review, you can review this PR first while hold the other (#6068).
Thank you :)

Commit:
kubelet: Add preparePod to check pod infra container.

Kill idle pod infra container when all containers exits.
Also fixed a few unit tests.

yifan-gu commented:
@vmarmol @yujuhong  Just rebased.

I moved the pod infra container logic into preparePod, but there is still some left:
https://github.com/GoogleCloudPlatform/kubernetes/pull/6169/files#diff-bf28da68f62a8df6e99e447c4351122dR1225
I think we can make an interface for the container runtime to pass the `containerToKeep` and `containersToStart` to it, then the runtime can determine to kill the pod infra container for us.

Also discussed with @jonboulle last week, I think such interface also works for `rkt` for now, which rkt will just kill/start the whole pod.

/cc @philips

yifan-gu commented:
Closing, since we are splitting this to #6608 and #6832 (though the later one is held for now)
